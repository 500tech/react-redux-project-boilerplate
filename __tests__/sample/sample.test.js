import React from 'react';
import { shallow, mount } from 'enzyme';
import { noop } from 'lodash/fp';

import { Sample, StyledPost } from 'sample/sample';

// mock the injection of the reducer into the application as we don't run
// tests with any reference to the store
jest.mock('sample/sample-replace-reducer', () => {});

const setup = ({ posts = [], isLoading = false, fetchPosts = noop } = {}) => ({
  posts,
  isLoading,
  fetchPosts
});

describe('<Sample />', () => {
  test('should render', () => {
    const props = setup();
    const component = shallow(<Sample {...props} />);

    expect(component).toMatchSnapshot();
  });

  test('should render loading when component is in loading state', () => {
    const props = setup({ isLoading: true });
    const component = shallow(<Sample {...props} />);

    expect(component).toMatchSnapshot();
  });

  test('should render 3 posts', () => {
    const posts = [
      { id: 1, title: 'first post', body: 'first post content' },
      { id: 2, title: 'second post', body: 'second post content' },
      { id: 3, title: 'third post', body: 'third post content' }
    ];
    const props = setup({ posts });
    const component = shallow(<Sample {...props} />);

    const postsCount = component.find(StyledPost).length;

    expect(postsCount).toEqual(3);
    expect(component).toMatchSnapshot();
  });

  test('should fetch posts when mounting', () => {
    const props = setup({ fetchPosts: jest.fn() });
    shallow(<Sample {...props} />);

    expect(props.fetchPosts).toHaveBeenCalled();
  });

  test('should fetch posts when clicking refresh button', () => {
    const props = setup({ fetchPosts: jest.fn() });
    const component = shallow(<Sample {...props} />);

    // We check that fetch posts have been called once already as it is run in componentDidMount
    expect(props.fetchPosts).toHaveBeenCalledTimes(1);

    component.find('img').simulate('click');

    expect(props.fetchPosts).toHaveBeenCalledTimes(2);
  });
});
