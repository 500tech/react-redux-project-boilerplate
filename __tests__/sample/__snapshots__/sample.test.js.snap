// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Sample /> should render 1`] = `
<sample__StyledContainer>
  <h1>
    <FormattedMessage
      id="homepage.title"
    />
  </h1>
  <h3>
    To get started, search your project for // TODO
    <br />
    This is a sample component that uses a sample action + reducer, it fetches posts from a remote server and displays them
  </h3>
  <img
    alt="refresh"
    onClick={[Function]}
    src="https://www.materialui.co/materialIcons/navigation/refresh_grey_192x192.png"
  />
  <h2>
    Posts from remote server
  </h2>
  <div />
</sample__StyledContainer>
`;

exports[`<Sample /> should render 3 posts 1`] = `
<sample__StyledContainer>
  <h1>
    <FormattedMessage
      id="homepage.title"
    />
  </h1>
  <h3>
    To get started, search your project for // TODO
    <br />
    This is a sample component that uses a sample action + reducer, it fetches posts from a remote server and displays them
  </h3>
  <img
    alt="refresh"
    onClick={[Function]}
    src="https://www.materialui.co/materialIcons/navigation/refresh_grey_192x192.png"
  />
  <h2>
    Posts from remote server
  </h2>
  <div>
    <sample__StyledPost
      key="1"
    >
      <h4>
        first post
      </h4>
      <p>
        first post content
      </p>
    </sample__StyledPost>
    <sample__StyledPost
      key="2"
    >
      <h4>
        second post
      </h4>
      <p>
        second post content
      </p>
    </sample__StyledPost>
    <sample__StyledPost
      key="3"
    >
      <h4>
        third post
      </h4>
      <p>
        third post content
      </p>
    </sample__StyledPost>
  </div>
</sample__StyledContainer>
`;

exports[`<Sample /> should render loading when component is in loading state 1`] = `
<sample__StyledContainer>
  <h1>
    <FormattedMessage
      id="homepage.title"
    />
  </h1>
  <h3>
    To get started, search your project for // TODO
    <br />
    This is a sample component that uses a sample action + reducer, it fetches posts from a remote server and displays them
  </h3>
  <img
    alt="refresh"
    onClick={[Function]}
    src="https://www.materialui.co/materialIcons/navigation/refresh_grey_192x192.png"
  />
  <h2>
    Posts from remote server
  </h2>
  <div>
    loading...
  </div>
</sample__StyledContainer>
`;
